File: src/main/java/com/pancisin/bookster/utils/UniqueSlugGenerator.java
package com.pancisin.bookster.utils;

import java.io.Serializable;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

import org.hibernate.HibernateException;
import org.hibernate.engine.spi.SessionImplementor;
import org.hibernate.id.IdentifierGenerator;

import com.github.slugify.Slugify;
import com.pancisin.bookster.models.Page;

public class UniqueSlugGenerator implements IdentifierGenerator {

	@Override
	public Serializable generate(SessionImplementor session, Object object) throws HibernateException {
		Page page = (Page) object;
		Slugify s = new Slugify();
		String slug = s.slugify(page.getName());
		
		Connection connection = session.connection();

		try {
			Statement statement = connection.createStatement();
			String query = "select count(id) as slug_count from pages WHERE slug LIKE '" + slug + "%'";

			ResultSet rs = statement.executeQuery(query);

			if (rs.next()) {
				slug = String.join("-", slug, String.valueOf(rs.getInt(1)));
			}
		} catch (SQLException ex) {
			ex.printStackTrace();
		}

		return slug;
	}

}

File: src/main/kotlin/com/pancisin/bookster/utils/UniqueSlugGenerator.kt
package com.pancisin.bookster.utils

import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Statement

import org.hibernate.Session
import org.hibernate.tuple.ValueGenerator

import com.github.slugify.Slugify
import com.pancisin.bookster.model.Page

class UniqueSlugGenerator : ValueGenerator<String> {

  override fun generateValue(session: Session, owner: Any): String? {
    val page = owner as Page
    var slug = Slugify().slugify(page.name)

    session.doWork { connection ->
      try {
        val query = "select count(id) as slug_count from pages WHERE slug LIKE '$slug%'"
        val rs = connection.createStatement().executeQuery(query)

        if (rs.next()) {
          val number = rs.getInt(1)

          if (number > 0) {
            slug = arrayOf(slug, rs.getInt(1).toString()).joinToString("-")
          }
        }
      } catch (ex: SQLException) {
        ex.printStackTrace()
      }
    }

    return slug
  }

}
